### ISSUE TITLE
Search for nearest neighbors with bounds on distances along different axes, not the eucledian distance

### ISSUE THREAD

**Issue by chhh** [davidmoten/rtree#42]
The problem is with this method:
```java
public Observable<Entry<T, S>> search(final Rectangle r, final double maxDistance) {
        return search(new Func1<Geometry, Boolean>() {
            @Override
            public Boolean call(Geometry g) {
                return g.distance(r) < maxDistance;
            }
        });
    }
```
which only accepts a single distance. 

It would be nice to have something like `public Observable<Entry<T, S>> search(final Rectangle r, final double maxDistX, double maxDistY)` or `public Observable<Entry<T, S>> search(final Rectangle r, final double[] maxDistance)`.

The use case: the dimensions of the axes are different, so using a single distance doesn't make much sense. Trying to bring both dimensions to the same scale is not so simple, plus I just know the constraints I want to put on each of those coordinates.

The fact that r-tree class is marked final and the constructors are private also doesn't help extending the code.

I know that I can run `RTree.search(Rectangle)` with the desired constraints, but there are cases where this can bring up a lot of unnecessary stuff (e.g. larger rectangles in the tree covering the same region) and cause unnecessary slowdows, while I only need to get a few neighbors of a certain point.

**Thanks for the great library!**

**Comment by davidmoten** [comment]
Glad you are finding the library useful!

If the dimensions of the axes are very different then that might have negative effects on the performance of the rtree because of bias to or away from an axis. I would normally suggest translating and scaling the coordinates in this case say so they are in the 0..1 range. 

Euclidean distance doesn't sound like it has a lot of meaning as you point out. I would just do a search with a rectangle `(x-xDistance, y-yDistance, x+xDistance, y+yDistance)` which seems to match your use case exactly. Am I missing something?

**Comment by chhh** [comment]
That's what I'm doing now (search with a rectangle), but it finds everything that **intersects** the query. In my case such a search might bring up a lot of unnecessary stuff (e.g. some large rectangles that are also in the tree), I thought that it could be much more efficient to do `nearest()` with small `k` because I only need the closest things to my query point.

After writing this I think I know what I'm missing -- a method like `search(Geometry g, boolean fullyEncloses)`. Instead of intersections I'd like to be able to find all things that are completely enclosed by the query. In case of arbitrary geometries that would be full enclosure of their bounding boxes.

I don't know if such queries are possible/easy to implement with an RTree. If not, then a `nearest()` with different constraints for coordinates might work.

**Comment by davidmoten** [comment]
To get the contains seach you would do this:

```java
RTree<Object, Rectangle> tree = ...;
List<Entry<Object,Rectangle>>  entries = 
  tree
    .search(rect)
    .filter(entry -> rect.contains(entry.geometry()))
    .toList().toBlocking().single(); 
```
This will be the most efficient way of doing it and any api additions would probably do this under the covers. Are you using `filter` with the returned observable as above?

**Comment by chhh** [comment]
No, I wasn't using `filter`, I thought that some direct serach solution would be faster. But since you say this is how you'd implement this type of query anyway, I just have to believe you :)

Thank you!

PS: I guess this can be added to the readme (how to perform `contains` queries)
